src/program.cpp

```diff
@@ -205,8 +208,70 @@ target_assignments program::get_target_assignments(const std::vector<target>& ta
 
 bool program::is_compiled() const { return not this->impl->target_name.empty(); }
 
+void program::compile(const std::vector<std::string>& targets,
+                      std::unordered_map<std::string, compile_options> compile_opt_map)
+{
+    // Gather all the target roots
+    std::unordered_multimap<std::string, module_ref> roots;
+    auto mods = this->get_modules();
+    for(auto* mod : mods)
+    {
+        for(const auto& ins : *mod)
+        {
+            if(ins.name() != "run_on_target")
+                continue;
+            auto v                       = ins.get_operator().to_value();
+            module_ref root              = ins.module_inputs().front();
+            std::string root_target_name = v.at("target").to<std::string>();
+            assert(contains(targets, root_target_name));
+            roots.insert({root_target_name, root});
+        }
+    }
+
+    auto trace = tracer{};
+    // TODO: Add tracer based on compile options
+    if(enabled(MIGRAPHX_TRACE_COMPILE{}))
+        trace = tracer{std::cout};
+
+    trace(*this);
+    trace();
+
+    // Run passes on each root target
+    for(const auto& t : targets)
+    {
+        auto root_target           = migraphx::make_target(t);
+        auto root_modules_range    = roots.equal_range(t);
+        this->impl->context_map[t] = root_target.get_context();
+        for(auto i = root_modules_range.first; i != root_modules_range.second; i++)
```

Comment:
Use `for(auto i:range(root_modules_range))`

Reply:
Done.